### ISSUE TITLE
[Core] Allow unblocking resource pool for more efficient io

### ISSUE THREAD

**Issue by dongreenberg** [ray-project/ray#33628]
### Description

When I know I'm about to do an io operation within a remote function, I'd like to be able to do something like:
```
with ray.free():
   # Some io work that doesn't consume the CPU or GPU resources I have allocated for this function
```
When I enter the block, the task or actor's resources are freed so other jobs can start, and when the block is done, it waits for the resources to be available before resuming. `free` could take a `resources` dict which only frees some resources but not others (e.g. I may be able to free some cpu, but I don't want to free gpu because I could end up waiting much longer before being able to restart). This would be a much easier way to achieve lots of hoop jumping people do through asyncio!

Another possible API could be:
```
@ray.remote(free=True)
def my_io_submethod():
    # ...

@ray.remote(num_cpus=4)
def my_outer_fn():
    # do some work
    ref = my_io_submethod.remote()  # unblocks the 4 cpus
    data = ray.get(ref) # Only returns when 4 cpus are available again

   # More work

```

### Use case

I have a scraping task I'm parallelizing with Ray. I have 20,000 urls I'd like to crawl, and they have a wide range of download times. I'm doing this on a 64 cpu box, calling scraping_fn.remote on all the urls, and then ray.get() on the list. When I first start the job it parallelizes nicely and cpu utilization is high, but it starts to gradually drop, and after an hour it's in the order of 10% of all cpus. I inspected top and found that a large number of the 64 ray processes are stuck waiting for the site to download, and they're blocking new jobs from starting. This could be an easy fix if I could wrap the download block inside my task function in `free`, allowing other jobs to start downloading while longer downloads are in progress. I don't see a way that I could do this user-side.

**Comment by clarng** [comment]
By default, when a task calls another task, we free the resources of the caller. Does that address your issue?

**Comment by dongreenberg** [comment]
Wow... I did not realize that. So if I'm within some task that has a bunch of resources (let's say 64 CPUs), and then I use ray to map over a bunch of 1 cpu tasks, Ray won't ensure that only 64 cpus are used total between my sub-tasks, it'll free up the whole pool when I enter a subtask and then pull each 1 cpu task from the global pool?

**Comment by jjyao** [comment]
@dongreenberg,

When a task blocks on `ray.get()`, it's CPU resource will be freed and will be reclaimed when `ray.get()` returns. This is the reason why Ray can run the following code

```
ray.init(num_cpus=1)

@ray.remtoe(num_cpus=1)
def child():
   return "hello"
 
@ray.remote(num_cpus=1)
def parent():
  return ray.get(child.remote()) // Temporarily release the cpu resource so that child task can acquire it and run.
```

**Comment by NicolasSchuler** [comment]
@jjyao Sorry for opening the thread, but can this also be done with actors? E.g. I called an actor and after the call I want the resources to be free, because the actor is doing nothing. How can one then reset the resources claimed by the actor without destroying him?
